### PR TITLE
fix(pie-icons-webc): DSW-1686 fix icon SSR issues

### PR DESCRIPTION
**Pull request by jamieomaguire**
## Describe your changes
- Fixes SSR issues relating to icons by:
  - Moving away from lifecycle methods such as `connectedCallback` to define sizes, which does not run on the server
  - Ensuring sizes are correctly applied to the SVG in the template and reactive
- Adds 3 icons stories to Storybook - 1 to show all icons, 1 for a regular icon, and another for the same icon as a large variant (uses a different SVG)
- Refactor the icon components to inherit from a base abstract icon class that houses all the lifecycle logic. This also slightly reduces the bundle size and makes the icon logic a little easier to read instead of having to read a template literal in the `build.js` script.
- Adds back snapshot files that were accidentally deleted a few months ago. We believe the tests on CI were cached which is why the test didn't fail previously. We have disabled caching the icon tests for now. These tests are needed to visual test the **React** exported versions of our icons. We can look to using percy for this in future.
- Added basic percy visual tests
---
"@justeattakeaway/pie-icons-webc": minor
---

[Changed] - Fix SSR related issues with web component icons and refactor TS classes to use a base abstract class for shared logic

---
"pie-storybook": minor
---

[Added] - pie-icons-webc story to display all icons

---
"pie-monorepo": minor
---

[Changed] - Updated CI so that package changes will trigger builds and tests of its dependents (such as a change in pie-icons-webc triggering visual tests for pie-icon-button

## Changes to bundle size
A happy side effect of the refactor. Not a huge change, but still something!

### Before
Screenshot 2024-04-24 at 17 08 23

### After
Screenshot 2024-04-24 at 17 08 17

## Author Checklist (complete before requesting a review)
- [x] I have performed a self-review of my code
- [x] I have added thorough tests where applicable (unit / component / visual)
- [x] I have reviewed the `PIE Storybook`/`PIE Docs` PR preview
- [x] I have reviewed visual test updates properly before approving
- [x] If changes will affect consumers of the package, I have created a changeset entry.
- [x] If a changeset file has been created, I have used the `/snapit` functionality to test my changes in a consuming application

### Reviewer 1
- [x] I have reviewed the `PIE Storybook`/`PIE Docs` PR preview
- [x] If there are visual test updates, I have reviewed them

### Reviewer 2
- [x] I have reviewed the `PIE Storybook`/`PIE Docs` PR preview
- [x] If there are visual test updates, I have reviewed them

### CODE REVIEW

File: packages/tools/pie-icons-webc/PieIconComponent.ts [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+    @state()
+    protected _svgHeight!: string | number;
+
+    protected abstract isLarge: boolean;
+    protected abstract name: string;
+    public abstract class: string;
+
+    firstUpdated (): void {
+        this.size = this.isLarge ? 32 : 'xs';
+        this._svgWidth = this.isLarge ? 32 : 16;
```
Comment (xander-marjoram) [inline packages/tools/pie-icons-webc/PieIconComponent.ts:44]: I reckon we could import and use the default sizes from pie-icons-configs here

Comment (xander-marjoram) [reply packages/tools/pie-icons-webc/PieIconComponent.ts:44]: https://github.com/justeattakeaway/pie/blob/main/packages/tools/pie-icons-configs/configs.js#L11-L13

Comment (jamieomaguire) [reply packages/tools/pie-icons-webc/PieIconComponent.ts:44]: great shout, updating now

File: apps/pie-storybook/types/StoryOptions.ts [hunk 2]
```diff
@@ -17,9 +17,11 @@ export type StoryBackgrounds = {
 
 type StorybookControls = {
     exclude?: string[],
-    include?: string[]
+    include?: string[],
 }
 export interface StoryOptions {
```
Comment (jamieomaguire) [inline apps/pie-storybook/types/StoryOptions.ts:22]: I think we can probably import types directly from Storybook, but I would prefer a separate ticket/PR to do this

File: packages/tools/pie-icons-webc/PieIconComponent.ts [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+    // "--icon-size-override" css var can be used in parent components to make sure the icon
+    // stays the correct size for cases where the icon is added by consumers via a slot,
+    // for example in Button, IconButton, Chip and Tag.
+    // Might be changed later to assigning a size prop to an icon slot
+    // (POC: https://github.com/justeattakeaway/pie/pull/1128)
+    static styles = css`
+        :host {
+            display: inline-block;
+            /* Inline SVGs can cause an additional ~4px height in their containers (such as this host).*/
+            line-height: 0;
```
Comment (jamieomaguire) [inline packages/tools/pie-icons-webc/PieIconComponent.ts:21]: Removing line height fixes this rogue `4px` issue

File: apps/pie-storybook/package.json [hunk 4]
```diff
@@ -34,6 +34,7 @@
     "dompurify": "3.0.11"
   },
   "devDependencies": {
+    "@justeattakeaway/pie-icons-configs": "4.5.1",
```
Comment (xander-marjoram) [inline apps/pie-storybook/package.json:37]: I'm not entirely sure how this works with storybook, but these are used inside the story, so should this go in `dependencies`?

Comment (jamieomaguire) [reply apps/pie-storybook/package.json:37]: I was unsure as well. its used to build the story so I thought dev deps but open to suggestions

Comment (xander-marjoram) [reply apps/pie-storybook/package.json:37]: I guess if it works it's fine for now 🤷🏻‍♂️

File: packages/tools/pie-icons-webc/PieIconComponent.ts [hunk 5]
```diff
@@ -0,0 +1,60 @@
+import {
+    LitElement, css, PropertyValues, type TemplateResult,
+} from 'lit';
+import { property, state } from 'lit/decorators.js';
+
+import {
+    getSvgProps, RegularIconSize, LargeIconSize,
```
Comment (xander-marjoram) [inline packages/tools/pie-icons-webc/PieIconComponent.ts:7]: I think we can do this?
```suggestion
    getSvgProps, type RegularIconSize, type LargeIconSize,
```

File: packages/tools/pie-icons-webc/PieIconComponent.ts [hunk 6]
```diff
@@ -0,0 +1,60 @@
+import {
+    LitElement, css, PropertyValues, type TemplateResult,
```
Comment (xander-marjoram) [inline packages/tools/pie-icons-webc/PieIconComponent.ts:2]: And this?
```suggestion
    LitElement, css, type PropertyValues, type TemplateResult,
```

File: packages/tools/pie-icons-webc/build.js [hunk 7]
```diff
@@ -14,78 +14,51 @@ const { icons } = pieIcons.default;
 
 const componentTemplate = (name, svg) => {
     const svgClasses = svg.match(/class="(.+?)"/)?.[1];
-
     const isLargeIcon = name.endsWith('Large');
     const sizeType = isLargeIcon ? 'LargeIconSize' : 'RegularIconSize';
+    // 32 or 'xs'
+    const defaultSize = isLargeIcon ? largeIconSizeDefault : `'${regularIconSizeDefault}'`;
```
Comment (xander-marjoram) [inline packages/tools/pie-icons-webc/build.js:20]: Would this not work?
```suggestion
    const defaultSize = isLargeIcon ? largeIconSizeDefault : regularIconSizeDefault;
```

File: packages/tools/pie-icons-webc/PieIconComponent.ts (hunk 6 above)
Comment (jamieomaguire) [reply packages/tools/pie-icons-webc/PieIconComponent.ts]: 🔥

File: packages/tools/pie-icons-webc/PieIconComponent.ts (hunk 5 above)
Comment (jamieomaguire) [reply packages/tools/pie-icons-webc/PieIconComponent.ts]: 🔥

File: packages/tools/pie-icons-webc/build.js (hunk 7 above)
Comment (jamieomaguire) [reply packages/tools/pie-icons-webc/build.js:20]: Screenshot 2024-04-26 at 17 22 25
Sadly it seems to not compile to a string when I try this approach 😢

**Review by xander-marjoram — APPROVED** [review]

**Review by siggerzz — APPROVED** [review]

**Pull request merged**